src/inmanta/deploy/scheduler.py

```diff
@@ -95,6 +96,71 @@ async def report_resource_state(
         """
 
 
+class TaskRunner:
+    def __init__(self, endpoint: str, scheduler: "ResourceScheduler"):
+        self.endpoint = endpoint
+        self.status = AgentStatus.STOPPED
+        self._scheduler = scheduler
+        self._task: typing.Optional[asyncio.Task[None]] = None
+
+    def start(self) -> None:
+        if self.status == AgentStatus.STOPPED:
+            self._task = asyncio.create_task(self.run())
+        else:
+            self.status = AgentStatus.STARTED
+
+    def stop(self) -> None:
+        self.status = AgentStatus.STOPPING
+
+    async def notify(self) -> None:
+        """
+        Method to notify the runner that something has changed in the DB. This method will fetch the new information
+        regarding the environment and the information related to the runner (agent). Depending on the desired state of the
+        agent, it will either stop / start the agent or do nothing
+        """
+        current_environment = await Environment.get_by_id(self._scheduler.environment)
+        assert current_environment
+        current_agent = await data.Agent.get(env=self._scheduler.environment, endpoint=self.endpoint)
+        if current_agent is None:
+            return
+        should_be_running = not (current_environment.halted or current_agent.paused)
+
+        match self.status:
+            case AgentStatus.STARTED if not should_be_running:
+                self.stop()
+                if current_environment.halted and self._task is not None:
+                    self._task.cancel("Environment is being halted")
```

Comment:
Not sure if that's actually useful, as this would never get called in time

Reply:
I would not do that, the run method is absolutely not sufficiently robust to handle the exception (e.g. `self.status = stopped` is not in a final block, the code below it may not handle it well either. )